File: Projects/rosalind/RSTR.java
import java.io.*;
import java.util.*;

public class RSTR {
	private static final Integer A = (int) 'A', C = (int) 'C', G = (int) 'G', T = (int) 'T';
	public static void main (String[] args) throws IOException {
		final Scanner in = new Scanner(System.in);
		final int n = in.nextInt();
		final double p = in.nextDouble();
		final String read = in.next();
		final int a = (int) read.chars().filter(A::equals).count();
		final int c = (int) read.chars().filter(C::equals).count();
		final int g = (int) read.chars().filter(G::equals).count();
		final int t = (int) read.chars().filter(T::equals).count();
		System.out.println(1 - Math.pow(1 - Math.pow(p/2, c+g) * Math.pow((1-p)/2, a+t), n));
	}
}

File: Projects/rosalind/RSTR.kt
package rosalind

import java.io.IOException
import java.util.*

object RSTR {
    private const val A = 'A'.toInt()
    private const val C = 'C'.toInt()
    private const val G = 'G'.toInt()
    private const val T = 'T'.toInt()
    @Throws(IOException::class)
    @JvmStatic
    fun main(args: Array<String>) {
        val stdin = Scanner(System.`in`)
        val n = stdin.nextInt()
        val p = stdin.nextDouble()
        val read = stdin.next()
        val a = read.chars().filter { obj: Int -> A.equals(obj) }.count().toInt()
        val c = read.chars().filter { obj: Int -> C.equals(obj) }.count().toInt()
        val g = read.chars().filter { obj: Int -> G.equals(obj) }.count().toInt()
        val t = read.chars().filter { obj: Int -> T.equals(obj) }.count().toInt()
        println(1 - Math.pow(1 - Math.pow(p / 2, c + g.toDouble()) * Math.pow((1 - p) / 2, a + t.toDouble()), n.toDouble()))
        stdin.close()
    }
}